fix #21370, regression in dynamic dispatch of complex constructors

JeffBezanson opened the pull request:
Here is a possible approach to #21370. With this change, the times get pretty close to 0.4 levels.

First, I added a fast-reject path to `sig_match_simple` to help avoid more calls to `jl_types_equal`.

More significantly, the problem is that we can't use hash lookup for `Type{T}`, so dispatching constructors was doing a really long linear lookup. We can get around that by using the type of the first argument as the primary key. That seems to work well in practice, but eventually we should come up with a scheme for splitting typemap levels on Type{T}. I believe we are most often looking for egal types, so one thing that might work would be to hash assuming we are looking for something egal, and fall back to linear lookup if that fails.

JeffBezanson committed:
fix #21370, regression in dynamic dispatch of complex constructors

JeffBezanson commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21377#issuecomment-294024867) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3d609abf7ff8ff195bb0603b52789fdbb265a703/1a7a2e2_vs_94417e4/report.md). cc @jrevels

JeffBezanson committed:
speed up BitArray `reshape`

This mitigates a slight performance regression due to the change
in constructor dispatch.

JeffBezanson commented:
Looks like there are a lot of nice speedups here. There was one significant regression, most likely because this change does not guarantee every constructor will dispatch faster. There were some lucky types at the beginning of the list in the old code. `BitArray{1}` seems to be one of those, which slowed down a benchmark that constructed a very large number of views. Fortunately the BItArray code could easily be improved by changing an `N` to a `Val{N}`.

As a side note, this means we are calling `reshape` to make a single-element view (`view(A, 1)`). I'm sure that can be avoided if we want to significantly improve this case.

tkelman commented:
@nanosoldier `runbenchmarks("array" || "io", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21377#issuecomment-294203851) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f1e2f91df6ae028daaf3014c2d9d38d90a3e6012/6b02922_vs_67e3b50/report.md). cc @jrevels

JeffBezanson commented:
LGTM.

cstjean commented:
> the problem is that we can't use hash lookup for Type{T}

I was looking at the dispatch code in `jl_lookup_generic_`, and I'd like to understand this comment better. From the overview

```julia
search order:
      check associative hash based on callsite address for leafsig match
      look at concrete signatures
      if there is an exact match, return it
      otherwise look for a matching generic signature
```
and the comment that types are hash-consed, I would think that "look at concrete signatures" could be done by looking up the hash of the type addresses in a hash-table. Why not?

AFAICT, this PR increased the speed of my very heterogeneous, very parametric-type-intensive benchmark by 4X, from 40 seconds to 10. :tada:

JeffBezanson commented:
The problem is that `Type{T}` matches any type equal to `T`, and type equality is not the same as structural equality, so hash-consing doesn't handle every case there.

cstjean commented:
> The problem is that Type{T} matches any type equal to T

Could you please give an example?

Would it be stupid of me to insert a hash-table lookup in-between the callsite check and the typemap, that would efficiently dispatch all non-`Type` methods by hashing the types of the arguments? Right now, dynamic dispatch on a parametric type instance is at least linear-time in the size of the arguments' types (because the typemap lookup traverses the whole type?), but given that the types are hash-consed, it sounds like constant-time should be achievable for non-Type dispatch, no?

I believe that you're suggesting something similar above, but I'm having a hard time understanding `typemap.c`

vtjnash commented:
There are many cases, some examples include:
- `T = Dict{K, V}` vs. `typealias DictVK{K, V} Dict{K, V}` (e.g. `Type{Dict}` needs to also match `Type{DictKV}`, since `Dict == DictVK` even though `Dict !== DictVK`)
- `Union{Tuple{Float32}, Tuple{Float64}}` vs. `Tuple{Union{Float32, Float64}}`. Again, these are `==`, but not `===`

When `isleaftype(T)`, we can assume that `==` is the same as `===`, in which case they do end up in a hash: https://github.com/JuliaLang/julia/blob/0bbccf1ecd5f6c2596e3fe91f463efd4446cefb6/src/typemap.c#L831 (actually, multiple hashes, since if no argument is a Type, it qualifies for being placed in an even simpler, faster hash https://github.com/JuliaLang/julia/blob/7cd0324e030dba49e90f327b539e03ee93025383/src/gf.c#L1841)

Otherwise, the typemap often needs to just scan all possible method matches, since they are only in a partial sort order in the table.

cstjean commented:
Thank you, that explanation is very helpful. The callsite cache is cool, but at `N_CALL_CACHE = 4096` I worry that for applications like mine that don't have a tight inner-loop, its benefits might be lost. 

> Otherwise, the typemap often needs to just scan all possible method matches, since they are only in a partial sort order in the table.

Wouldn't it be worth caching the result of that linear search in a permanent data structure? Eg. for

```julia
f(x, y) = A
f{T}(x, y::Type{T}) = B

# called with
f(5, 6.0)
f(5, "hey")
f(10, Dict)
f(11, DictKV)
f(12, Complex)
```

I propose adding a hash-table cache in-between the `callsite` cache and the typemap, with entries:

```julia
(&Int, &Float64) => method_instance_1             # & for address-of (i.e. egal comparison)
(&Int, &String) => method_instance_2
(&Int, &DataType) => ((2, ), sub_hash_table_1)    # sub-key is the value of 2nd argument

where sub_hash_table_1 has entries:

(&Dict,) => method_instance_3
(&DictKV,) => method_instance_3 
(&Complex,) => method_instance_4
```

The entries will be either filled in at the `instantiate the generic method` step, or by the `typemap` look-up (eg. handling the `DictKV` case). Then any call with previously-seen type combinations would be dispatched in at most two hash-table look-ups. Is there anything obviously wrong with this proposal?

vtjnash commented:
I don't usually see that cache end up with more than a couple of hundred entries. There just aren't that many generic call sites. As long as you don't perform dispatch using types, it's pretty likely you'll end up hitting this case, regardless of how your code is structured.

That hash-table already exists (see my first link above), but many of the objects you are putting in there aren't valid for it. For instance, it can't contain Type{DictKV} or DataType, since we don't dispatch those objects using the `egal` predicate.

JeffBezanson commented:
@vtjnash is right. Of course improvements are possible, but they are a bit difficult. For example, we could try dispatching using `egal` first, and fall back to slower lookup only when that fails.

cstjean commented:
Ok, thank you for the feedback.
